mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -69,41 +69,49 @@ impl S3CrtClient {
 
             let callback = review_callback.clone();
 
-            let mut options = S3CrtClientInner::new_meta_request_options(message, S3Operation::PutObject);
+            let mut options = message.into_options(S3Operation::PutObject);
             options.send_using_async_writes(true);
             options.on_upload_review(move |review| callback.invoke(review));
             options.part_size(self.inner.write_part_size as u64);
 
-            let on_mpu_created_sender = Mutex::new(Some(mpu_created_sender));
-
-            self.inner.make_simple_http_request_from_options(
+            let on_mpu_created_sender = Arc::new(Mutex::new(Some(mpu_created_sender)));
+            let on_failure_sender = on_mpu_created_sender.clone();
+            self.inner.make_meta_request_with_callbacks(
                 options,
                 span,
                 move |metrics| {
                     if metrics.request_type() == RequestType::CreateMultipartUpload && !metrics.error().is_err() {
-                        // Signal that a CreateMultipartUpload completed successfully (unless the meta-request had already failed).
                         if let Some(sender) = on_mpu_created_sender.lock().unwrap().take() {
-                            _ = sender.send(());
+                            _ = sender.send(Ok(()));
                         }
                     }
                 },
-                |_| None,
                 on_headers,
+                |_, _| {},
+                |_| None,
+                move |result| {
+                    if let Some(sender) = on_failure_sender.lock().unwrap().take() {
```

Comment:
Maybe using an atomic bool instead of `Arc<Mutex<Option<...>>>` could be simpler here

Reply:
Not sure what you have in mind.

Another approach I considered was to introduce an "event" channel like in `get_object`, but for `MpuCreated` and `Result`. However, it would require more changes to `S3PutObjectRequest` and we are considering removing the wait for CreateMultipartUpload anyway, so I decided to go for the less intrusive change.

Reply:
I was thinking of having only one channel for all results and keeping track of errors in a `AtomicBool`, but I think that way the initial `CreateMultipartUploadFailed` error would be emitted once the caller first polls this method rather than initially from this method.